contracts/Witch.sol

```diff
@@ -9,79 +9,141 @@ import "./math/WMul.sol";
+    function buy(bytes12 vaultId, uint128 base, uint128 min)
+        public
+        returns (uint256 ink)
+    {
         DataTypes.Balances memory balances_ = cauldron.balances(vaultId);
+        DataTypes.Vault memory vault_ = cauldron.vaults(vaultId);
+        Auction memory auction_ = auctions[vaultId];
+        (uint256 duration_, uint256 initialOffer_, uint256 dust_) = (duration, initialOffer, dust);
 
         require (balances_.art > 0, "Nothing to buy");                                      // Cheapest way of failing gracefully if given a non existing vault
-        uint256 elapsed = uint32(block.timestamp) - cauldron.auctions(vaultId);           // Auctions will malfunction on the 7th of February 2106, at 06:28:16 GMT, we should replace this contract before then.
-        uint256 price;
+        uint256 art = cauldron.debtFromBase(vault_.seriesId, base);
         {
-            // Price of a collateral unit, in underlying, at the present moment, for a given vault
-            //
-            //                ink                     min(auction, elapsed)
-            // price = 1 / (------- * (p + (1 - p) * -----------------------))
-            //                art                          auction
-            (uint256 auctionTime_, uint256 initialProportion_) = (auctionTime, initialProportion);
-            uint256 term1 = uint256(balances_.ink).wdiv(balances_.art);
-            uint256 dividend2 = auctionTime_ < elapsed ? auctionTime_ : elapsed;
-            uint256 divisor2 = auctionTime_;
-            uint256 term2 = initialProportion_ + (1e18 - initialProportion_).wmul(dividend2.wdiv(divisor2));
-            price = uint256(1e18).wdiv(term1.wmul(term2));
+            uint256 elapsed = uint32(block.timestamp) - auction_.start;                      // Auctions will malfunction on the 7th of February 2106, at 06:28:16 GMT, we should replace this contract before then.
+            uint256 price = inkPrice(balances_, initialOffer_, duration_, elapsed);
+            ink = uint256(art).wdivup(price);                                                    // Calculate collateral to sell. Using divdrup stops rounding from leaving 1 stray wei in vaults.
+            require (ink >= min, "Not enough bought");
+            require (ink == balances_.ink || balances_.ink - ink >= dust_, "Leaves dust");
         }
-        uint256 ink = uint256(art).wdivup(price);                                                    // Calculate collateral to sell. Using divdrup stops rounding from leaving 1 stray wei in vaults.
-        require (ink >= min, "Not enough bought");
 
-        ladle.settle(vaultId, msg.sender, ink.u128(), art);                                        // Move the assets
+        cauldron.slurp(vaultId, ink.u128(), art.u128());                                            // Remove debt and collateral from the vault
+        ladle.settle(vaultId, msg.sender, ink.u128(), base);                                        // Move the assets
```

Comment:
It is somewhat concerning to me that the Ladle is a dependency of Witch. What happens if we need to upgrade Ladle, can it cause issues with already operating auctions? If we upgrade Ladle, what is the likelihood that we need to update Witch?

Reply:
Right now, if the Ladle is replaced, the Witch must be replaced as well. The process to let the old Witch finish ongoing auctions would be as follows:

1. New Ladle and new Witch are deployed, but not permissioned yet (so that everything fits in a transaction).
2. New Ladle and new Witch are configured.
3. New Ladle and new Witch are permissioned, allowing them to start operations immediately. Old Witch is removed  from `Cauldron.grab` permissions to stop new auctions being initiated with it. The old Ladle is removed from all Cauldron permissions, which still allows the old Witch to finish ongoing auctions using `ladle.settle`.
4. Once all auctions in the old Witch are finished, the `ladle.settle` permissions are removed. The old Ladle is removed from the permissions of all the Joins.

I will remove the Ladle dependency from the Witch, though. The dependency is there because I was concerned of governance errors in setting up new assets (which would need to be set up on both the Ladle and the Witch). Using the Wand, though, the possibility of error is removed.